README.md
<p align="center"><img src="https://avatars0.githubusercontent.com/u/28169525?s=200&v=4" height="100"/></p>
<h1 align="center"><a href="https://captainfact.io">CaptainFact.io</a></h1>
<p align="center"><a href="https://discord.gg/2Qd7hMz" title="Discord"><img src="https://discordapp.com/api/guilds/416782744748687361/widget.png" alt="Discord"></a>
<a href="https://twitter.com/CaptainFact_io" title="Twitter"><img src="https://img.shields.io/twitter/follow/CaptainFact_io.svg?style=social&label=Follow"></a>
<a href="https://opencollective.com/captainfact_io" title="Backers on Open Collective"><img src="https://opencollective.com/captainfact_io/backers/badge.svg"></a>
<a href="./LICENSE"><img src="https://img.shields.io/github/license/CaptainFact/captain-fact-api.svg" alt="AGPL3"></a>
<a href="https://travis-ci.com/CaptainFact/captain-fact-api"><img src="https://travis-ci.com/CaptainFact/captain-fact-api.svg?branch=staging"></a>
<a href='https://coveralls.io/github/CaptainFact/captain-fact-api?branch=staging'><img src='https://coveralls.io/repos/github/CaptainFact/captain-fact-api/badge.svg?branch=staging' alt='Coverage Status' /></a>
</p>
<hr/>
<p align="center">
<a href="https://opencollective.com/captainfact_io/donate" target="_blank">
  <img src="https://opencollective.com/captainfact_io/donate/button@2x.png?color=white" width=300 />
</a>
</p>
<hr/>
<br/>

# Install & Run

## Prerequisites

You need to install Elixir. We recommand using [asdf](https://github.com/asdf-vm/asdf#setup).
Check their documentation on how to install it, then run `asdf install` from
root `captain-fact-api` folder.

## Start DB

Create / launch a postgres instance on your local machine. If you have
docker installed, you can use the pre-seed postgres docker image:

`docker run -d --name cf_dev_db -p 5432:5432 captainfact/dev-db:latest`

## Start API

- `mix deps.get` --> Get dependencies
- `mix ecto.migrate` --> Migrate DB
- `iex -S mix` --> Start project

Following services will be started:

- [localhost:4000](http://localhost:4000) - REST API
- [localhost:4001](https://localhost:4001) - REST API (https)
- [localhost:4002](http://localhost:4002) - GraphQL API
- [localhost:4003](https://localhost:4003) - GraphQL API (https)
- [localhost:4004](http://localhost:4004) - Atom feed

You can also see all the mail sent by going to http://localhost:4000/_dev/mail

## Other useful commands

- `mix test` --> Run tests
- `mix test.watch` --> Run tests watcher
- `mix format` --> Format code
- `mix ecto.gen.migration [migration_name]` --> Generate migration

# Project architecture

This application is organized as an [umbrella project](https://elixir-lang.org/getting-started/mix-otp/dependencies-and-umbrella-apps.html) which allows us to divide CaptainFact API into small apps.

```
.
├── apps
│   ├── cf => Core functions as a library. **Not deployed**
│   ├── cf_atom_feed => Atom feed.
│   ├── cf_graphql => GraphQL API (public).
│   ├── cf_jobs => Jobs.
│   ├── cf_opengraph => An app that generate opengraph tags.
│   ├── cf_rest_api => REST/WS API (private).
│   └── db => DB repository and schemas **Not deployed**
│       ├── lib
│       │   ├── db
│       │   ├── db_schema => Contains all the schemas (Video, Speaker, Comment...etc)
│       │   ├── db_type => Special types (SpeakerPicture...etc)
│       │   └── db_utils => Some utils functions
│       └── priv
│           └── repo/migrations => All DB migrations files
├── README.md => You're reading it right now. Are you?
├── rel => Release configs & tools
│   ├── commands => Commands that will be available to run on the release (seed DB...etc)
│   ├── hooks => Some hooks for automatically run commands when release run.
│   ├── runtime_config => Runtime configurations for all apps.
│   └── config.exs => Releases configuration.
```

# Linked projects

- [Community discussions and documentation](https://github.com/CaptainFact/captain-fact/)
- [Frontend](https://github.com/CaptainFact/captain-fact-frontend)
- [Extension](https://github.com/CaptainFact/captain-fact-extension)
- [Overlay injector](https://github.com/CaptainFact/captain-fact-overlay-injector)


CHANGELOG.md
# Change Log

## [v1.0.2](https://github.com/CaptainFact/captain-fact-api/tree/v1.0.2) (2019-04-01)
[Full Changelog](https://github.com/CaptainFact/captain-fact-api/compare/v1.0.1...v1.0.2)

**Fixed bugs:**

- GenServer terminating \(MatchError\): no match of right hand side value: {:error, :cannot\_create\_table} [\#135](https://github.com/CaptainFact/captain-fact-api/issues/135)
- GenServer terminating \(MatchError\): no match of right hand side value: {:error, :cannot\_create\_table} [\#134](https://github.com/CaptainFact/captain-fact-api/issues/134)

**Merged pull requests:**

- Improve CI tests and release configs [\#141](https://github.com/CaptainFact/captain-fact-api/pull/141) ([Betree](https://github.com/Betree))
- Edit speaker enhancements [\#138](https://github.com/CaptainFact/captain-fact-api/pull/138) ([Betree](https://github.com/Betree))
- fix\(RuntimeConfig\): Basic auth password [\#137](https://github.com/CaptainFact/captain-fact-api/pull/137) ([Betree](https://github.com/Betree))
- chore\(deps\): bump timex from 3.3.0 to 3.5.0 [\#113](https://github.com/CaptainFact/captain-fact-api/pull/113) ([dependabot[bot]](https://github.com/apps/dependabot))

## [v1.0.1](https://github.com/CaptainFact/captain-fact-api/tree/v1.0.1) (2019-03-25)
[Full Changelog](https://github.com/CaptainFact/captain-fact-api/compare/v1.0.0...v1.0.1)

**Fixed bugs:**

- DBConnection.ConnectionError: connection not available because of disconnection [\#132](https://github.com/CaptainFact/captain-fact-api/issues/132)
- GenServer terminating \(DBConnection.ConnectionError\): connection not available because of disconnection [\#131](https://github.com/CaptainFact/captain-fact-api/issues/131)
- DBConnection.ConnectionError: connection not available because of disconnection [\#130](https://github.com/CaptainFact/captain-fact-api/issues/130)
- GenServer terminating \(ArgumentError\): argument error [\#127](https://github.com/CaptainFact/captain-fact-api/issues/127)

**Merged pull requests:**

- Release 1.0.1 [\#133](https://github.com/CaptainFact/captain-fact-api/pull/133) ([Betree](https://github.com/Betree))
- fix\(SubscriptionsMatcher\): Don't return canceled subscriptions [\#129](https://github.com/CaptainFact/captain-fact-api/pull/129) ([Betree](https://github.com/Betree))
- Release 1.0.0 [\#126](https://github.com/CaptainFact/captain-fact-api/pull/126) ([Betree](https://github.com/Betree))

## [v1.0.0](https://github.com/CaptainFact/captain-fact-api/tree/v1.0.0) (2019-03-19)
[Full Changelog](https://github.com/CaptainFact/captain-fact-api/compare/v0.9.3...v1.0.0)

**Fixed bugs:**

- GenServer terminating: exited in: GenServer.call\(CF.RestApi.Presence, {:list, "video\_debate:4P7J"}, 5000\)
    \*\* \(EXIT\) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application i [\#110](https://github.com/CaptainFact/captain-fact-api/issues/110)

**Merged pull requests:**

- Notifications [\#60](https://github.com/CaptainFact/captain-fact-api/pull/60) ([Betree](https://github.com/Betree))

## [v0.9.3](https://github.com/CaptainFact/captain-fact-api/tree/v0.9.3) (2019-01-05)
[Full Changelog](https://github.com/CaptainFact/captain-fact-api/compare/v0.9.2...v0.9.3)

**Fixed bugs:**

- CF.Accounts.UserPermissions.PermissionsError: unauthorized [\#92](https://github.com/CaptainFact/captain-fact-api/issues/92)

**Closed issues:**

- Add API key to users [\#50](https://github.com/CaptainFact/captain-fact-api/issues/50)
- Ability to export verification results [\#10](https://github.com/CaptainFact/captain-fact-api/issues/10)
- User account deletion : soft delete [\#8](https://github.com/CaptainFact/captain-fact-api/issues/8)
- Add a cleanup job to remove expired password reset requests [\#7](https://github.com/CaptainFact/captain-fact-api/issues/7)
- \[New achievements\] Hit machine, fact's tamer, you made a point [\#5](https://github.com/CaptainFact/captain-fact-api/issues/5)

**Merged pull requests:**

- Release 0.9.3 [\#109](https://github.com/CaptainFact/captain-fact-api/pull/109) ([Betree](https://github.com/Betree))
- Videos: add unlisted and get all added by user [\#107](https://github.com/CaptainFact/captain-fact-api/pull/107) ([Betree](https://github.com/Betree))
- chore\(ActivityLog\): Show user banned actions [\#105](https://github.com/CaptainFact/captain-fact-api/pull/105) ([Betree](https://github.com/Betree))
- chore\(deps\): bump floki from 0.20.3 to 0.20.4 [\#104](https://github.com/CaptainFact/captain-fact-api/pull/104) ([dependabot[bot]](https://github.com/apps/dependabot))
- chore\(deps-dev\): bump mix\_test\_watch from 0.8.0 to 0.9.0 [\#103](https://github.com/CaptainFact/captain-fact-api/pull/103) ([dependabot[bot]](https://github.com/apps/dependabot))
- chore\(deps-dev\): bump ex\_machina from 2.2.0 to 2.2.2 [\#102](https://github.com/CaptainFact/captain-fact-api/pull/102) ([dependabot[bot]](https://github.com/apps/dependabot))
- chore\(deps-dev\): bump bypass from 0.8.1 to 0.9.0 [\#99](https://github.com/CaptainFact/captain-fact-api/pull/99) ([dependabot[bot]](https://github.com/apps/dependabot))
- feat\(Limitations\): Check actions on a 15 minutes period instead of 24h [\#98](https://github.com/CaptainFact/captain-fact-api/pull/98) ([Betree](https://github.com/Betree))
- chore\(deps\): bump distillery from 2.0.11 to 2.0.12 [\#95](https://github.com/CaptainFact/captain-fact-api/pull/95) ([dependabot[bot]](https://github.com/apps/dependabot))
- chore\(deps\): bump absinthe\_plug from 1.4.5 to 1.4.6 [\#93](https://github.com/CaptainFact/captain-fact-api/pull/93) ([dependabot[bot]](https://github.com/apps/dependabot))
- chore\(deps\): bump basic\_auth from 2.2.2 to 2.2.4 [\#91](https://github.com/CaptainFact/captain-fact-api/pull/91) ([dependabot[bot]](https://github.com/apps/dependabot))

config/dev.exs
use Mix.Config


dev_secret = "8C6FsJwjV11d+1WPUIbkEH6gB/VavJrcXWoPLujgpclfxjkLkoNFSjVU9XfeNm6s"

# General config
config :captain_fact,
  frontend_url: "http://localhost:3333",
  cors_origins: [
    # Frontend Dev
    "http://localhost:3333", "https://localhost:3333",
    "http://192.168.20.6:3333",
    # Overlay injector tester
    "http://localhost:3342",
    # Misc
    "http://localhost", "https://localhost",
    # Extension
    "chrome-extension://fnnhlmbnlbgomamcolcpgncflofhjckm",
  ]

# For development, we disable any cache and enable
# debugging and code reloading.
config :captain_fact, CaptainFactWeb.Endpoint,
  secret_key_base: dev_secret,
  debug_errors: false,
  code_reloader: true,
  check_origin: false,
  http: [port: 4000],
  force_ssl: false,
  https: [
    port: 4001,
    otp_app: :captain_fact,
    keyfile: "priv/keys/privkey.pem",
    certfile: "priv/keys/fullchain.pem"
  ]

# Guardian + Ueberauth

config :guardian, Guardian,
  secret_key: dev_secret

config :ueberauth, Ueberauth.Strategy.Facebook.OAuth,
  client_id: "506726596325615",
  client_secret: "4b320056746b8e57144c889f3baf0424"

# Do not include metadata nor timestamps in development logs
config :logger, :console, format: "[$level] $message\n"

# Set a higher stacktrace during development. Avoid configuring such
# in production as building large stacktraces may be expensive.
config :phoenix, :stacktrace_depth, 20

# Configure file upload
config :arc, storage: Arc.Storage.Local

# Configure your database
config :captain_fact, CaptainFact.Repo,
  username: "postgres",
  password: "postgres",
  database: "captain_fact_dev",
  hostname: "localhost",
  pool_size: 10

# Mails
config :captain_fact, CaptainFact.Mailer, adapter: Bamboo.LocalAdapter

# Env / Secrets are above everything else
# Weave loads config from env or secret files
config :weave,
       file_directories: ["priv/secrets"],
       loaders: [Weave.Loaders.File, Weave.Loaders.Environment]